crypto/x509/x509_lu.c

```diff
@@ -323,33 +328,47 @@ int X509_STORE_CTX_get_by_subject(const X509_STORE_CTX *vs,
 
     if (!X509_STORE_lock(store))
         return 0;
-
     tmp = X509_OBJECT_retrieve_by_subject(store->objs, type, name);
     X509_STORE_unlock(store);
 
     if (tmp == NULL || type == X509_LU_CRL) {
         for (i = 0; i < sk_X509_LOOKUP_num(store->get_cert_methods); i++) {
             lu = sk_X509_LOOKUP_value(store->get_cert_methods, i);
-            j = X509_LOOKUP_by_subject_ex(lu, type, name, &stmp, vs->libctx,
-                                          vs->propq);
-            if (j) {
+            if (lu->skip)
+                continue;
+            if (lu->method == NULL)
+                return -1;
+            j = X509_LOOKUP_by_subject_ex(lu, type, name, &stmp,
+                                          vs->libctx, vs->propq);
+            if (j != 0) { /* non-zero value is considered success here */
                 tmp = &stmp;
                 break;
             }
         }
         if (tmp == NULL)
             return 0;
     }
+    if (tmp->type != type)
+        return -1;
 
-    if (!X509_OBJECT_up_ref_count(tmp))
-        return 0;
-
-    ret->type = tmp->type;
-    ret->data.ptr = tmp->data.ptr;
+    if (ret != NULL) {
```

Comment:
Is this NULL check correct... What is the point of this call if ret is passed in..

Reply:
> Is this NULL check correct...

It is correct, but not strictly needed, as it was not present before in  `X509_STORE_CTX_get_by_subject()`,
and the calls done within `x509_lu.c` check the `ret` argument they pass in beforehand.

>  What is the point of this call if ret is passed in..

The `ret` parameter is passed in such that the function is able to return this way the components of the found object.
A bit unusual, but that's the way the function was designed, and I did not touch this design.